src/main/java/org/tikv/txn/TxnKVClient.java

```diff
@@ -1,5 +1,363 @@
+            return false;
+        }
+    }
+
+    /**
+     * Request for batch rollback on TiKV, retry operation should be deal with Caller
+     * @param backOffer
+     * @param keys
+     * @param startTs
+     * @param regionId
+     * @return
+     */
+    public ClientRPCResult batchRollbackReq(BackOffer backOffer, byte[][] keys, long startTs, long regionId) {
+        ClientRPCResult result = new ClientRPCResult(true, false, null);
+        List<ByteString> byteList = Lists.newArrayList();
+        for(byte[] key : keys) {
+            byteList.add(ByteString.copyFrom(key));
+        }
+        try {
+            TiRegion region = regionManager.getRegionById(regionId);
+            RegionStoreClient client = clientBuilder.build(region);
+            //send request
+            client.batchRollback(backOffer, byteList, startTs);
+        } catch (final Exception e) {
+            result.setSuccess(false);
+            result.setRetry(e instanceof RegionException);//mark retryable, region error, should retry prewrite again
+            result.setError(e.getMessage());
+        }
+        return result;
+    }
+
+    /**
+     * Get value of key from TiKV
+     * @param key
+     * @return
+     */
+    public byte[] get(byte[] key) {
+        ByteString byteKey = ByteString.copyFrom(key);
+        BackOffer bo = ConcreteBackOffer.newGetBackOff();
+        long version = 0;
+        ByteString value = null;
+        try {
+            RegionStoreClient client = clientBuilder.build(byteKey);
+            version =  getTimestamp().getVersion();
+            value = client.get(bo, byteKey, version);
+        } catch (final TiKVException | StatusRuntimeException e) {
+            LOG.error("Get process error, key={}, version={}", new String(key), version);
+        }
+
+        return value != null ? value.toByteArray() : new byte[0];
+    }
+
+    /**
+     * Put a new key-value pair to TiKV
+     * @param key
+     * @param value
+     * @return
+     */
+    public boolean put(byte[] key, byte[] value) {
+        boolean putResult = false;
+        ByteString byteKey = ByteString.copyFrom(key);
+        ByteString byteValue = ByteString.copyFrom(value);
+        BackOffer bo = ConcreteBackOffer.newCustomBackOff(BackOffer.prewriteMaxBackoff);
+        List<Kvrpcpb.Mutation> mutations = Lists.newArrayList(
+                Kvrpcpb.Mutation.newBuilder()
+                        .setKey(byteKey).setValue(byteValue).setOp(Kvrpcpb.Op.Put)
+                        .build()
+        );
+        long lockTTL = 2000;
+        long startTS;
+        TiRegion region = regionManager.getRegionByKey(byteKey);
+        boolean prewrite = false;
+        while(true) {
+            try {
+                startTS = this.getTimestamp().getVersion();
+                ClientRPCResult prewriteResp = this.prewrite(bo, mutations, key, lockTTL, startTS, region.getId());
```

Comment:
Does it seem we can only get a retriable error when meeting `RegionException`? So could we use the same `startTS` when we meet the `RegionException`?

Reply:
You means that only exception which contains Write and write conflict error should refresh txn startTs ? RegionException meets when server encounters region slipt , we needn't change startTs in this situation.
Is it right ?

Reply:
> Does it seem we can only get a retriable error when meeting `RegionException`? So could we use the same `startTS` when we meet the `RegionException`?

You means that only exception which contains Write and write conflict error should refresh txn startTs ? RegionException meets when server encounters region slipt , we needn't change startTs in this situation.
Is it right ?